tests/performance/metrics/metrics.go

```diff
@@ -0,0 +1,212 @@
+const (
+	Seconds    = "seconds"
+	MegaBytes  = "MB"
+	Percentage = "percentage"
+)
+
+func (v *MetricType) String() string {
+	switch *v {
+	case Pending:
+		return "Pending"
+	case Scheduling:
+		return "Scheduling"
+	case Scheduled:
+		return "Scheduled"
+	case Running:
+		return "Running"
+	case Terminated:
+		return "Terminated"
+
+	case VMIPodMemory:
+		return "VMIPodMemory"
+	case VMIPodCPU:
+		return "VMIPodCPU"
+
+	case VirtOperatorCPU:
+		return "VirtOperatorCPU"
+	case VirtOperatorMemory:
+		return "VirtOperatorMemory"
+	case VirtControllerCPU:
+		return "VirtControllerCPU"
+	case VirtControllerMemory:
+		return "VirtControllerMemory"
+	case VirtApiCPU:
+		return "VirtApiCPU"
+	case VirtApiMemory:
+		return "VirtApiMemory"
+	case VirtApiReqDuration:
+		return "VirtApiReqDuration"
+	case VirtHandlerCPU:
+		return "VirtHandlerCPU"
+	case VirtHandlerMemory:
+		return "VirtHandlerMemory"
+
+	default:
+		return "VmPhaseUnset"
+	}
+}
+
+func GetMetricType(metric string) MetricType {
+	switch metric {
+	case "Pending":
+		return Pending
+	case "Scheduling":
+		return Scheduling
+	case "Scheduled":
+		return Scheduled
+	case "Running":
+		return Running
+	case "Terminated":
+		return Terminated
+
+	case "VMIPodMemory":
+		return VMIPodMemory
+	case "VMIPodCPU":
+		return VMIPodCPU
+
+	case "VirtOperatorCPU":
+		return VirtOperatorCPU
+	case "VirtOperatorMemory":
+		return VirtOperatorMemory
+	case "VirtControllerCPU":
+		return VirtControllerCPU
+	case "VirtControllerMemory":
+		return VirtControllerMemory
+	case "VirtApiCPU":
+		return VirtApiCPU
+	case "VirtApiMemory":
+		return VirtApiMemory
+	case "VirtApiReqDuration":
+		return VirtApiReqDuration
+	case "VirtHandlerCPU":
+		return VirtHandlerCPU
+	case "VirtHandlerMemory":
+		return VirtHandlerMemory
+
+	default:
+		return VmPhaseUnset
+	}
+}
+
+// Metric is a struct for metrics for performance tests, it can be latencies or resource usage.
+type Metric struct {
+	Perc50 float64 `yaml:"perc50"`
+	Perc90 float64 `yaml:"perc90"`
+	Perc99 float64 `yaml:"perc99"`
+	Unit   Unit    `yaml:"unit,omitempty"`
+}
+
+// VMMetricData encapsulates vm startup latency information.
+type VMMetricData struct {
+	// Name of the vm
+	Name string
+	// Node this vm was running on
+	Node string
+	// Metric value
+	Value float64
+	// Metric unit
+	Unit Unit
+	// Type can be latency (e.g., for each VM creation phases), or resource (e.g., of resource usage)
+	Type int
+}
+
+// Slice is an array of VMMetricData which encapsulates vm startup latency information.
+type Slice []VMMetricData
+
+func (a Slice) Len() int           { return len(a) }
+func (a Slice) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a Slice) Less(i, j int) bool { return a[i].Value < a[j].Value }
+
+func (a Slice) Sort() Slice { sort.Sort(a); return a }
+
+// ExtractMetrics returns latency metrics for each percentile(50th, 90th and 99th).
+func ExtractMetrics(data []VMMetricData) Metric {
+	metrics := Metric{}
```

Comment:
Can this leverage the work going on in #5766 ? 5766 calculated the transitions times so you can scrape the endpoint.

Reply:
Well, yes and no... The problem to be strictly dependent on the Prometheus metric is that the test can only run if Prometheus is deployed. Which is not done by default in `kubevirtci`, although, I have enabled the optinal deployment of Prometheus in `kubevirtci` in [#608](https://github.com/kubevirt/kubevirtci/pull/608).

Prometheus could potentially be a strong requirement for us, but, I intended to have this test as flexible as possible so that everyone could execute it into their environment.

Note that, for the perspective of our CI/CD system, the metrics from #5766 will be very valuable to visualize the tests later in a Grafana dashboard.

Reply:
I guess that is something we have to discuss further. It would be weird to use one set of metrics to decide if we meet certain goals but we use different data to visualize the data.